runtime-modules/bounty/src/lib.rs

```diff
@@ -1151,102 +1325,235 @@ decl_module! {
+
+            let bounty_creator_manager = BountyActorManager::<T>::get_bounty_actor_manager(
+                bounty.creation_params.creator.clone(),
+            )?;
+
+            //
+            // == MUTATION SAFE ==
+            //
 
-            // Fire an event.
-            Self::deposit_event(RawEvent::WorkEntrantFundsWithdrawn(bounty_id, entry_id, member_id));
+            bounty_oracle_manager.transfer_funds_from_bounty_account(bounty_id, oracle_reward);
 
-            // Remove the bounty in case of the last withdrawal operation.
-            if Self::withdrawal_completed(&current_bounty_stage, &bounty_id) {
-                Self::remove_bounty(&bounty_id);
+            <Bounties<T>>::mutate(bounty_id, |bounty| {
+                bounty.has_unpaid_oracle_reward = false;
+            });
+
+            Self::deposit_event(RawEvent::BountyOracleRewardWithdrawal(
+                bounty_id,
+                bounty.creation_params.oracle.clone(),
+                oracle_reward
+            ));
+
+            if Self::has_no_contributions_and_no_work_entries(&bounty_id) {
+                Self::remove_bounty(
+                    &bounty_id,
+                    &bounty,
+                    &bounty_creator_manager
+                );
             }
         }
     }
 }
 
 impl<T: Trait> Module<T> {
+    fn withdraw_funding_mutation(
+        bounty_id: &T::BountyId,
+        bounty: &Bounty<T>,
+        funder: BountyActor<MemberId<T>>,
+        bounty_funder_manager: &BountyActorManager<T>,
+        funding: Contribution<T>,
+    ) {
+        let cherry_fraction = Self::get_cherry_fraction_for_member(&bounty, funding.amount);
+
+        let withdrawal_amount = funding
+            .total_bloat_bond_and_funding()
+            .saturating_add(cherry_fraction);
+
+        bounty_funder_manager.transfer_funds_from_bounty_account(*bounty_id, withdrawal_amount);
+
+        <BountyContributions<T>>::remove(&bounty_id, &funder);
+
+        Self::deposit_event(RawEvent::FunderStateBloatBondWithdrawn(
+            *bounty_id,
+            funder.clone(),
+            T::FunderStateBloatBondAmount::get(),
+        ));
+
+        Self::deposit_event(RawEvent::BountyFundingWithdrawal(*bounty_id, funder));
+    }
+
+    fn withdraw_funding_state_bloat_bond_mutation(
+        bounty_id: &T::BountyId,
+        funder: BountyActor<MemberId<T>>,
+        bounty_funder_manager: &BountyActorManager<T>,
+        funding: Contribution<T>,
+    ) {
+        bounty_funder_manager
+            .transfer_funds_from_bounty_account(*bounty_id, funding.funder_state_bloat_bond_amount);
+
+        //Remove contribution from
+        <BountyContributions<T>>::remove(&bounty_id, &funder);
+
+        Self::deposit_event(RawEvent::FunderStateBloatBondWithdrawn(
+            *bounty_id,
+            funder,
+            T::FunderStateBloatBondAmount::get(),
+        ));
+    }
```

Comment:
This does the logic that the previous version of withdraw_funder_state_bloat_bond_amount did